File: HGTD/Oscilloscope/plot_2d.c
#include<fstream>
#include <iostream>
#include <vector>

using namespace std;
const TString plotPath = "Plots/";

void plot_2d(TString dataPath)
{   
    gErrorIgnoreLevel = kError;

    gStyle->SetLegendBorderSize(0);
    gStyle->SetLegendTextSize(.03);

    TCanvas *c1 = new TCanvas("c1","A Simple 2D Graph",200,100,750,500);
    c1->SetRightMargin(0.32);


    TString txtFile;
    const char  *entry;
    void *dataDir = gSystem->OpenDirectory(dataPath);

    // double x[5] = {0};
    // double y[5] = {0};
    // double z;

    int i, j, n;
    double x,y,z,A,x0,y0,xrange[2]={50,300},yrange[2]={0,50},xbins,ybins;
    
    TH2 *h_2d;

    xbins = 150;
    ybins = 50;
    // h_2d = new TH2D("plotname","plotname",xbins,xrange[0],xrange[1],ybins,yrange[0],yrange[1]);
    h_2d = new TH2D("plotname","plotname",200,0,500,10,0,10*2.5*15);
    
    while((entry = (char *)gSystem->GetDirEntry(dataDir)))
    {
        txtFile = entry;
        
        if(txtFile.EndsWith(".txt") || txtFile.EndsWith(".csv") )
        {
            cout<<txtFile<<endl;
            vector<double>  x_list,y_list,z_list;
            
            TString txtFileFullPath = dataPath + txtFile;

            TTree *tree = new TTree("graph","graph");
            // tree->ReadFile(txtFileFullPath, "x/D:y/D", ',');
            tree->ReadFile(txtFileFullPath,"x/D:y/D:z/D:A/D",',');
            tree->SetBranchAddress("x",&x);
            tree->SetBranchAddress("y",&y);
            tree->SetBranchAddress("A",&z);

            tree->GetEntry(0);

            x0 = x;
            y0 = y;

            for(n = 0; n < tree->GetEntries(); n++)
            {
                tree->GetEntry(n);
                
                // if(x != x0 || y != y0)
                if(x != xbins)
                {
                x_list.push_back(x-x0);
                y_list.push_back((y-y0)/15);
                z_list.push_back(z*1000+15);
                // z_list.push_back(-z*1000);
                // cout<<x<<","<<y<<","<<-z<<endl;

                // x0 = x;
                xbins = x;
                }
            }

            // h_2d = new TH2D("plotname","plotname",x_list.size(),xrange[0],xrange[1],y_list.size(),yrange[0],yrange[1]);

            for(i = 0; i < x_list.size(); i++)
            {
                // for(j = 0; j < sqrt(y_list.size()); j++)
                // {
                    if(i%10 == 0 ) cout<<x_list[i]<<","<<y_list[i]<<","<<z_list[i]<<endl;
                    // if(z_list[i] > 37)  
                    // A = 0.6;
                    // else    A = 0.1;
                    h_2d->SetBinContent(x_list[i],y_list[i],z_list[i]);
                    // h_2d->SetBinContent(int(x_list[i]/4),y_list[i],A);
                    // if(z_list[i*sqrt(y_list.size())+j] > 2.5) h_2d->SetBinContent(j+1,i+1,z_list[i*sqrt(y_list.size())+j]);
                    // if(z_list[i*sqrt(y_list.size())+j] > 2.5) h_2d->SetBinContent(i+1,j+1,1);
                    // else h_2d->SetBinContent(i+1,j+1,0);
                    // if(i/100 == 0 )  cout<<z_list[i]<<endl;
                // }
            }
            // cout<<i<<endl;
        }
    }

    // Int_t MyPalette[100];
    // Double_t r[]    = {0., 0.0, 1.0, 1.0, 1.0};
    // Double_t g[]    = {0., 0.0, 0.0, 1.0, 1.0};
    // Double_t b[]    = {0., 1.0, 0.0, 0.0, 1.0};
    // Double_t stop[] = {0., .25, .50, .75, 1.0};
    // Int_t FI = TColor::CreateGradientColorTable(5, stop, r, g, b, 50);
    // for (int i=0;i<50;i++) MyPalette[i] = FI+i;
    
    // gStyle->SetOptStat();

    h_2d->GetXaxis()->SetNdivisions(5, kTRUE);
    h_2d->GetYaxis()->SetNdivisions(5, kTRUE);
    
    // h_2d->GetXaxis()->SetRangeUser(0, 300);
    // h_2d->GetYaxis()->SetRangeUser(0.5, 4);
    // h_2d->GetZaxis()->SetRangeUser(0, 50);
    // h_2d->GetZaxis()->SetRangeUser(0, 1);
    // h_2d->SetContour(99);

    // h_2d->GetXaxis()->SetTitle("Bias [V]");
    // h_2d->GetYaxis()->SetTitle("Charge [fC]");
    // h_2d->GetZaxis()->SetTitle("Time Resolution [ps]");

    h_2d->GetXaxis()->SetTitle("X [um]");
    h_2d->GetYaxis()->SetTitle("Y [um]");
    h_2d->GetZaxis()->SetTitle("Ampl [a.u.]");
    
    h_2d->Draw("COLZ");
    // h_2d->Draw("CONT");

    c1->SaveAs(plotPath + txtFile.ReplaceAll(".txt","") + "_2d.png");
    c1->SaveAs(plotPath + txtFile.ReplaceAll(".txt","") + "_2d.svg");
    c1->SaveAs(plotPath + txtFile.ReplaceAll(".txt","") + "_2d.pdf");
    // c1->SaveAs(plotPath + "VCT_2d.png");
    // c1->SaveAs(plotPath + "VCT_2d.svg");
    // c1->SaveAs(plotPath + "VCT_2d.pdf");
}

File: HGTD/Oscilloscope/plot_2d.py
import ROOT
import os
import sys
from array import array

ROOT.gErrorIgnoreLevel = ROOT.kError
ROOT.gROOT.SetBatch(ROOT.kTRUE)
ROOT.gStyle.SetOptTitle(0)
ROOT.gStyle.SetOptStat(0)

c1 = ROOT.TCanvas("c1", "canvas", 1000, 800)

c1.SetLeftMargin(0.15)
# c1.SetRightMargin(0.05)
c1.SetTopMargin(0.05)
c1.SetBottomMargin(0.15)
c1.SetRightMargin(0.20)


indir = "Data/"
plotPath = "Plots/"

x = sys.argv[1]
f_list = os.listdir(indir)

# x = []
# y = []
i = 0
step = 2.5  #   2.5 um

for f in f_list:
    A = [[],[],[]]
    # if not "csv" in f and not "txt" in f:    continue
    if not "txt" in f:    continue
    if "charge" in f:    continue

    out = f.replace(".csv","").replace(".txt","")

    # f_out = open(plotPath + out + "_" + x + ".csv","a")
    f_out = open(plotPath + out + ".csv","a")
    f_out.seek(0)
    f_out.truncate()

    f = os.path.join(indir,f)
    print f
    with open(f,"r") as text:
        content = text.read().splitlines()
    
    x0 = -1
    y0 = -1
    # i = 0
    for p in content:
        # i+=1
        # if i > 10: break
        p = p.split(",")
        # if int(p[0]) > x0:    x0 = int(p[0])
        # if int(p[1]) > y0:    y0 = int(p[1])
        # A.append( float(p[8]) )
        # if int(p[0][3:6]) > x0:    x0 = int(p[0][3:6])
        # if int(p[0][7:]) > y0:    y0 = int(p[0][7:])
        A[0].append( int(p[0][3:6]) )
        A[1].append( int(p[0][7:]) )
        A[2].append( float(p[1]) )

        if int(p[0][3:6]) == int(x):
        # if int(p[0][7:]) == int(x):
        # if int(p[0][7:]) == 2000:
            # if int(p[0][7:]) < 2079 and int(p[0][7:]) > 2040:
            f_out.write( p[0][3:6] + "," + p[0][7:] + "," + p[1] + "\n")

    
    # x0 = x0 - int(content[0][3:6])
    # y0 = y0 - int(content[0][7:11])

    print(min(A[0]),min(A[1]),len(A[2]))

    h_2d = ROOT.TH2D("2d","2d", max(A[0])-min(A[0]), 0, (max(A[0])-min(A[0]))*step, max(A[1])-min(A[1]), 0, (max(A[1])-min(A[1]))*step )

    for i in range(len(A[0])):
        # if i > 10: break
        # print A[0][i]-min(A[0]), A[1][i]-min(A[1]), A[2][i]
        h_2d.SetBinContent( A[0][i]-min(A[0]), A[1][i]-min(A[1]), A[2][i] )

    # for i in range(x0):
    #     for j in range(y0):
    #         h_2d.SetBinContent(i,j,A[i+j/y0*x0])

    h_2d.GetXaxis().SetTitle("X [um]")
    h_2d.GetYaxis().SetTitle("Y [um]")
    h_2d.GetZaxis().SetTitle("Ampl [a.u.]")
    # h_2d.GetXaxis().SetNdivisions(505, ROOT.kTRUE)
    # h_2d.SetMarkerSize(0.2)
    h_2d.Draw("COLZ")

    # c1.SaveAs(plotPath + out + "_2d.png")
    # c1.SaveAs(plotPath + out + "_2d.svg")
    c1.Clear()
    
####    fraction calculation
f_out = open("fra_"+ str(x) +".csv","a")
f_out.seek(0)
f_out.truncate()

Pos = [[],[]]
Var = [[],[],[],[]]

for i in range(4):
    f = plotPath + "amplitude_" + str(i+1) + ".csv"
    print f
    with open(f,"r") as text:
        content = text.read().splitlines()

        for var in content:
            var = var.split(",")
            Var[i].append(float(var[2]))
            if i == 0:
                Pos[0].append(var[0])
                Pos[1].append(var[1])

for i,var in enumerate(Var[0]):
    # print Pos[0][i],Pos[1][i],Var[0][i]
    fra1 = (Var[0][i]-Var[1][i])/(Var[0][i]+Var[1][i])
    # fra2 = (Var[2][i]-Var[3][i])/(Var[2][i]+Var[3][i])
    fra2 = (Var[0][i]+Var[2][i]-Var[1][i]-Var[3][i])/(Var[0][i]+Var[2][i]+Var[1][i]+Var[3][i])
    # fra2 = (Var[0][i]-Var[2][i]+Var[1][i]-Var[3][i])/(Var[0][i]+Var[2][i]+Var[1][i]+Var[3][i])
    f_out.write( Pos[0][i] + "," + Pos[1][i] + "," + str(fra1) + "," + str(fra2) + "\n" )

f_out.close()